### PR TITLE
S3t11 backend validation for new tasks

### PR DESCRIPTION
**Pull request by AlisonYoon**

### CODE REVIEW

File: src/Classes/Entities/EventEntity.php [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 class EventEntity
 {
+    protected $id;
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: Why did you add id? The id is generated by the database, not supplied by the user.

File: src/Classes/Entities/EventEntity.php [hunk 2]
```diff
@@ -13,20 +14,170 @@ class EventEntity
...
+    }
+
+    /**
+     * Will sanitise all the fields for an applicant.
+     */
+    private function sanitiseData()
+    {
+        $this->id = (int) $this->id;
+        $this->name = $this->sanitiseString($this->name);
+        $this->name = self::validateLength($this->name, 20);
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: Any reason to limit length to 20 rather than 255?

File: src/Classes/Entities/EventEntity.php [hunk 3]
```diff
@@ -13,20 +14,170 @@ class EventEntity
...
+        $this->category = (int)$this->category;
+        $this->location = $this->sanitiseString($this->location);
+        $this->location = self::validateLength($this->location, 255);
+        $this->date = $this->sanitiseString($this->date);
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: The dates and times should use some regex to make sure that they look like dates and times in the format that mySQL will be expecting e.g. `'2019-10-02'`, `'16:55'`

File: src/Classes/Entities/EventEntity.php (hunk 1 above)
Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: Yeah I was confused if I should add this or not, seeing that they did in the ApplicantEntity, I did the same. I will remove it.

File: src/Classes/Entities/EventEntity.php (hunk 2 above)
Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: valid point

File: src/Classes/Entities/EventEntity.php [hunk 4]
```diff
@@ -13,20 +14,170 @@ class EventEntity
...
+        $this->sanitiseData();
+    }
+
+    /**
+     * Will sanitise all the fields for an applicant.
+     */
+    private function sanitiseData()
+    {
+        $this->id = (int) $this->id;
+        $this->name = $this->sanitiseString($this->name);
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: All of these string fields should probably have length checks

File: src/Classes/Entities/EventEntity.php [hunk 5]
```diff
@@ -13,20 +14,170 @@ class EventEntity
...
+        $this->name = self::validateLength($this->name, 20);
+        $this->category = (int)$this->category;
+        $this->location = $this->sanitiseString($this->location);
+        $this->location = self::validateLength($this->location, 255);
+        $this->date = $this->sanitiseString($this->date);
+        $this->startTime = $this->sanitiseString($this->startTime);
+        $this->endTime = $this->sanitiseString($this->endTime);
+        $this->notes = $this->sanitiseString($this->notes);
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: This should probably have a length check too.

Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: exist and length check maybe?

File: src/Classes/Entities/EventEntity.php [hunk 6]
```diff
@@ -13,20 +14,170 @@ class EventEntity
...
         $this->date = ($this->date ?? $date);
         $this->startTime = ($this->startTime ?? $startTime);
         $this->endTime = ($this->endTime ?? $endTime);
         $this->notes = ($this->notes ?? $notes);
+
+        $this->sanitiseData();
+    }
+
+    /**
+     * Will sanitise all the fields for an applicant.
```
Comment (lizsharp) [inline src/Classes/Entities/EventEntity.php]: for an event not an applicant

File: src/Classes/Entities/EventEntity.php [hunk 7]
```diff
@@ -13,20 +14,170 @@ class EventEntity
...
+        $this->location = $this->sanitiseString($this->location);
+        $this->location = self::validateLength($this->location, 255);
+        $this->date = $this->sanitiseString($this->date);
+        $this->startTime = $this->sanitiseString($this->startTime);
+        $this->endTime = $this->sanitiseString($this->endTime);
+        $this->notes = $this->sanitiseString($this->notes);
+    }
+
+    /**(
+     * Sanitise as a string in the applicant table as data.
```
Comment (lizsharp) [inline src/Classes/Entities/EventEntity.php]: event/applicant

File: src/Classes/Entities/EventEntity.php [hunk 8]
```diff
@@ -13,20 +14,170 @@ class EventEntity
...
+     *
+     * @param string $eventData
+     *
+     * @return string, which will return the applicant data.
```
Comment (lizsharp) [inline src/Classes/Entities/EventEntity.php]: event/applicant

File: src/Classes/Entities/EventEntity.php [hunk 9]
```diff
@@ -13,20 +14,170 @@ class EventEntity
...
+            return $eventData;
+        } else {
+            throw new \Exception('An input string does not exist or is too long');
+        }
+    }
+
+    /**
+     * Validate that a string is not empty and is within length allowed, throws an error if not
+     *
+     * @param string $hiringPartnerData
```
Comment (lizsharp) [inline src/Classes/Entities/EventEntity.php]: eventdata not hiringpartnerdata

File: src/Classes/Entities/EventEntity.php (hunk 5 above)
Comment (mallowmew) [reply src/Classes/Entities/EventEntity.php]: Notes doesn't need to exist. Other fields do.

File: src/Classes/Entities/EventEntity.php [hunk 10]
```diff
@@ -28,5 +28,173 @@ public function __construct(
...
+     */
+    public function sanitiseDate($date)
+    {
+        if (!preg_match('/([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))/', $date)) {
+            return false;
+        } else {
+            return $this->sanitiseString($this->date);
+        }
+    }
+
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: doc block?

File: src/Classes/Entities/EventEntity.php [hunk 11]
```diff
@@ -28,5 +28,173 @@ public function __construct(
...
+        $this->sanitiseData();
+    }
+
+    /**
+     * Will sanitise all the fields for an event.
+     */
+    private function sanitiseData()
+    {
+        $this->name = $this->sanitiseString($this->name);
+        $this->name = self::validateLength($this->name, 255);
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: Required fields should probably validate that they exist and are within length

File: src/Classes/Entities/EventEntity.php [hunk 12]
```diff
@@ -28,5 +28,173 @@ public function __construct(
...
+        $this->category = (int)$this->category;
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: Would it be possible to check if the category exists in the database?

File: src/Classes/Entities/EventEntity.php [hunk 13]
```diff
@@ -28,5 +28,173 @@ public function __construct(
...
+    }
+
+    /**
+     * Sanitise as a date in the event table as data.
+     *
+     * @param $date
+     *
+     * @return bool|string
+     */
+    public function sanitiseDate($date)
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: This and the time functions I would say are validating rather than sanitising - sanitisation normally means 'making the same data cleaner' while these are checking that the data matches an expected pattern.

File: src/Classes/Entities/EventEntity.php (hunk 10 above)
Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: ok

File: src/Classes/Entities/EventEntity.php (hunk 11 above)
Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: ok

File: src/Classes/Entities/EventEntity.php (hunk 13 above)
Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: ok

File: src/Classes/Entities/EventEntity.php (hunk 12 above)
Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: yes but not in a clean way

File: src/Classes/Entities/EventEntity.php [hunk 14]
```diff
@@ -28,5 +28,179 @@ public function __construct(
...
+     * Sanitise as a date in the event table as data.
+     *
+     * @param $date
+     *
+     * @return bool|string
+     */
+    public function validateDate($date)
+    {
+        if (!preg_match('/([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))/', $date)) {
+            return false;
```
Comment (mallowmew) [inline src/Classes/Entities/EventEntity.php]: This will assign a value of false to the date property of the entity - what then happens to that value of false on the way to the database?

File: src/Classes/Entities/EventEntity.php [hunk 15]
```diff
@@ -28,5 +28,179 @@ public function __construct(
...
+    }
+
+    /**
+     * Sanitise as a date in the event table as data.
+     *
+     * @param $date
+     *
+     * @return bool|string
+     */
+    public function validateDate($date)
```
Comment (FelixLetheren) [inline src/Classes/Entities/EventEntity.php]: Could $date be typehinted?

File: src/Classes/Entities/EventEntity.php [hunk 16]
```diff
@@ -28,5 +28,179 @@ public function __construct(
...
+     *
+     * @param string $eventData
+     *
+     * @return string, which will return the event data.
+     */
+    public function sanitiseString($eventData)
+    {
+        return filter_var($eventData, FILTER_SANITIZE_STRING);
+    }
+
```
Comment (MotNosniktaw) [inline src/Classes/Entities/EventEntity.php]: why use a named function that only contains a builtin function?

Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: this named function was used in other entities so I just did the same how it was done before.

File: src/Classes/Entities/EventEntity.php (hunk 14 above)
Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: changed to throw Exception

File: src/Classes/Entities/EventEntity.php [hunk 17]
```diff
@@ -28,5 +28,179 @@ public function __construct(
...
+    {
+        if (!preg_match('/((1[0-2]|0?[1-9]):([0-5][0-9]))/', $time)
+        ) {
+            throw new \Exception('Please enter correct time');
+        } else {
+            if ($time == $this->startTime) {
+                return $this->sanitiseString($this->startTime);
+            } elseif ($time == $this->endTime) {
+                return $this->sanitiseString($this->endTime);
+            }
```
Comment (FelixLetheren) [inline src/Classes/Entities/EventEntity.php]: This doesn't cover every eventuality as there isn't any return if none of these requirments are met.

File: src/Classes/Entities/EventEntity.php [hunk 18]
```diff
@@ -84,7 +84,7 @@ public function validateTime(string $time)
 
     public function validateStartEndTime($startTime, $endTime)
     {
-        if($startTime >= $endTime) {
+        if ($startTime >= $endTime) {
             throw new \Exception('End time should be later than Start time');
```
Comment (FelixLetheren) [inline src/Classes/Entities/EventEntity.php]: Should the S of start be capatalised?

File: src/Classes/Entities/EventEntity.php [hunk 19]
```diff
@@ -28,5 +28,185 @@ public function __construct(
...
+        $this->endTime = $this->validateTime($this->endTime);
+        $this->validateStartEndTime($this->startTIme, $this->endTime);
+        $this->notes = $this->sanitiseString($this->notes);
+        $this->notes = self::validateLength($this->notes, 255);
+    }
+
+    /**
+     * Sanitise as a date in the event table as data.
+     *
+     * @param $date
```
Comment (FelixLetheren) [inline src/Classes/Entities/EventEntity.php]: doc bock not up to date with type hinting.

File: src/Classes/Entities/EventEntity.php [hunk 20]
```diff
@@ -28,5 +28,185 @@ public function __construct(
...
+            throw new \Exception('Please enter correct date');
+        } else {
+            return $date;
+        }
+    }
+
+    /**
+     * Sanitise as a time in the event table as data.
+     *
+     * @param $time
```
Comment (FelixLetheren) [inline src/Classes/Entities/EventEntity.php]: doc block not up to date with type hinting

File: src/Classes/Entities/EventEntity.php [hunk 21]
```diff
@@ -28,5 +28,185 @@ public function __construct(
...
+    public function validateTime(string $time)
+    {
+        if (!preg_match('/((1[0-2]|0?[1-9]):([0-5][0-9]))/', $time)
+        ) {
+            throw new \Exception('Please enter correct time');
+        } else {
+            return $time;
+        }
+    }
+
```
Comment (FelixLetheren) [inline src/Classes/Entities/EventEntity.php]: missing docblock

File: src/Classes/Entities/EventEntity.php [hunk 22]
```diff
@@ -28,5 +28,185 @@ public function __construct(
...
+    public function validateStartEndTime($startTime, $endTime)
+    {
+        if ($startTime >= $endTime) {
+            throw new \Exception('End time should be later than start time');
+        } else {
+            return true;
+        }
+    }
+
+    /**(
```
Comment (FelixLetheren) [inline src/Classes/Entities/EventEntity.php]: what's this little ( doing over here? Cheeky little fella.

Comment (AlisonYoon) [reply src/Classes/Entities/EventEntity.php]: respect

**Comment by FelixLetheren** [comment]
**wowee**
